src/ucp/rndv/proto_rndv.c

```diff
@@ -631,15 +637,180 @@ void ucp_proto_rndv_bulk_query(const ucp_proto_query_params_t *params,
+    size_t rkey_length;
+    uint64_t remote_address;
+    size_t remote_size, accumulate_size;
+    ucs_status_t status;
+
+    p = ucp_proto_rndv_unpack_iov_count(p, &iov_count);
+    if (iov_count == 0) {
+        return UCS_ERR_NO_MESSAGE;
+    }
+
+    req->send.rndv.rma_array = ucs_malloc(
+            iov_count * sizeof(*req->send.rndv.rma_array), "proto_rndv");
+    if (req->send.rndv.rma_array == NULL) {
+        ucs_error("failed to allocate rma array");
+        status = UCS_ERR_NO_MEMORY;
+        goto finish;
+    }
+
+    accumulate_size = 0;
+    for (iov_index = 0; iov_index != iov_count; ++iov_index) {
+        p = ucp_proto_rndv_unpack_iov_rkey_buffer(p, &rkey_buffer, &rkey_length,
+                                                  &remote_address,
+                                                  &remote_size);
+        ucp_proto_rndv_check_rkey_length(remote_address, remote_size, "rts");
+        ucp_proto_rndv_check_rkey_length((uint64_t)rkey_buffer, rkey_length,
+                                         "rts");
+
+        status = ucp_ep_rkey_unpack_internal(
+                ep, rkey_buffer, rkey_length, ep_config->key.reachable_md_map,
+                ep_config->rndv.proto_rndv_rkey_skip_mds, &rkey);
+        if (status != UCS_OK) {
+            goto destroy_rkey;
+        }
+        accumulate_size                                    += remote_size;
+        req->send.rndv.rma_array[iov_index].accumulate_size = accumulate_size;
+        req->send.rndv.rma_array[iov_index].remote_address  = remote_address;
+        req->send.rndv.rma_array[iov_index].remote_size     = remote_size;
+        req->send.rndv.rma_array[iov_index].rkey            = rkey;
+        ucs_debug("index=%u rkey_size=%lu address=%p size=%lu", iov_index,
+                  rkey_length, (void*)remote_address, remote_size);
+    }
+    req->send.rndv.rma_count = iov_count;
+    req->send.rndv.rma_index = 0;
+    status                   = UCS_OK;
+    goto finish;
+
+destroy_rkey:
+    req->send.rndv.rma_count = iov_index;
+    ucp_proto_rndv_iov_rkeys_destroy(req);
+finish:
+    ucs_assertv(p <= p_end, "bounds error 0x%" PRIx64 ">0x%" PRIx64,
+                (uint64_t)p, (uint64_t)p_end);
+    return status;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t ucp_proto_rndv_unpack_common_rkey(
+        const void *buffer, size_t length, uint64_t remote_address,
+        size_t remote_size, ucp_ep_h ep, ucp_request_t *req)
+{
+    ucp_ep_config_t *ep_config = ucp_ep_config(ep);
+
+    if (length == 0) {
+        goto unpack_nothing;
+    }
+
+    ucs_assert(buffer != NULL);
+
+    if (remote_address != 0) {
+        ucp_proto_rndv_check_rkey_length(remote_address, length, "rts");
+        req->send.rndv.remote_address = remote_address;
+        return ucp_proto_rndv_unpack_contig_rkey(buffer, length, ep, ep_config,
+                                                 req);
+    }
+
+    if (req->send.state.dt_iter.dt_class != UCP_DATATYPE_CONTIG ||
```

Comment:
How do we reach `ucp_proto_rndv_unpack_iov_rkeys` if this `if` is `true` for `UCP_DATATYPE_IOV`?

Reply:
The sender side is support IOV, while the receiver side is not.
If we need IOV support in the receiver side, it will have additional work for the RNDV-GET proto.